### PR TITLE
Add ``__new__`` and ``__call__`` to``ObjectModel`` and ``ClassModel``

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

/CC @jacobtylerwalls 

Part of https://github.com/PyCQA/astroid/pull/1519.

This was what I was talking about in https://github.com/PyCQA/astroid/pull/1519#discussion_r890105397. ``__new__`` and ``__call__`` are part of the models for `object` and `classes`. Thus, instead of adding them as ``attr___ne__`` I think it makes sense to try and define them.

One thing I wonder is whether this should indeed be properties. But looking at other methods I think they should? See for example:
https://github.com/PyCQA/astroid/blob/01db68b1b37fdfa4d7899fa1ba71015b33efb34b/astroid/interpreter/objectmodel.py#L288-L289

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

**Comment by DanielNoord** [comment]
> Thanks for pushing forward on this. I haven't found time to do a review yet, but I did quickly check it against [PyCQA/pylint#6402](https://github.com/PyCQA/pylint/pull/6402) and this doesn't pass its test case. Was that expected?

Yeah! Ultimately we should merge #1519 as well.

However, since `__new__` and `__init__` are used so ubiquitously I thought there would be benefits to defining them rather than making them `attr___ne__`. I think for the other dunders most tests will automatically pass if we define them on `ObjectModel` instead of `FunctionModel`. (Although I haven't checked).

File: astroid/bases.py [hunk 1]
```diff
@@ -32,6 +32,7 @@
 objectmodel = lazy_import("interpreter.objectmodel")
 helpers = lazy_import("helpers")
 manager = lazy_import("manager")
+nodes = lazy_import("nodes")
```
Comment (jacobtylerwalls) [inline astroid/bases.py:35]: I didn't find this to be necessary earlier today in another PR. WDYT?

File: astroid/interpreter/objectmodel.py [hunk 2]
```diff
@@ -118,6 +118,19 @@ def lookup(self, name):
             return getattr(self, IMPL_PREFIX + name)
         raise AttributeInferenceError(target=self._instance, attribute=name)
 
+    @property
+    def attr___new__(self):
+        """Calling cls.__new__(cls) on an object returns an instance of that object.
+
+        Instance is either an instance or a class definition of the instance to be
+        created.
+        """
```
Comment (jacobtylerwalls) [inline astroid/interpreter/objectmodel.py:127]: I'm wondering if this obviates most of `UnboundMethod.infer_call_result`? Or no?

Comment (DanielNoord) [reply astroid/interpreter/objectmodel.py:127]:
```python
import astroid

mod = astroid.parse(
    """
class A:
    ...

A().__new__(A)
A.__new__(A)
    """
)
list(mod.body[-1].value.infer())
list(mod.body[-2].value.infer())
```

With:
```diff
diff --git a/astroid/bases.py b/astroid/bases.py
index e41964341..dd1e9d2c1 100644
--- a/astroid/bases.py
+++ b/astroid/bases.py
@@ -401,6 +401,7 @@ class UnboundMethod(Proxy):
             qname = self._proxied.parent.frame(future=True).qname()
             # Avoid checking builtins.type: _infer_type_new_call() does more validation
             if qname.startswith("builtins.") and qname != "builtins.type":
+                print("CALLED")
                 return self._infer_builtin_new(caller, context)
         return self._proxied.infer_call_result(caller, context)
 
diff --git a/astroid/interpreter/objectmodel.py b/astroid/interpreter/objectmodel.py
index d120895aa..65be429ce 100644
--- a/astroid/interpreter/objectmodel.py
+++ b/astroid/interpreter/objectmodel.py
@@ -127,6 +127,7 @@ class ObjectModel:
         """
         # TODO: Use isinstance instead of try ... except after _instance has typing
         try:
+            print("CALLED")
             return self._instance._proxied.instantiate_class()
         except AttributeError:
             return self._instance.instantiate_class()
```

Shows that this is used for bound methods. Both are being called in other instances I think.

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 3]
```diff
@@ -2298,7 +2298,12 @@ def infer_call_result(self, caller, context=None):
         try:
             metaclass = self.metaclass(context=context)
             if metaclass is not None:
-                dunder_call = next(metaclass.igetattr("__call__", context))
+                # Only get __call__ if it's defined locally for the metaclass.
+                # Otherwise we will find ObjectModel.__call__ which will
+                # return an instance of the metaclass. Instantiating the class is
+                # handled later.
+                if "__call__" in metaclass.locals:
+                    dunder_call = next(metaclass.igetattr("__call__", context))
```
Comment (jacobtylerwalls) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:2306]: Should there be a default value for next?

**Review by jacobtylerwalls — COMMENTED** [review]
Looks really good! One last Q

File: astroid/nodes/scoped_nodes/scoped_nodes.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/nodes/scoped_nodes/scoped_nodes.py:2306]: Caught by `StopIteration` on L2307.

Comment (jacobtylerwalls) [reply astroid/nodes/scoped_nodes/scoped_nodes.py:2306]: Thanks, didn't see it!

**Review by jacobtylerwalls — APPROVED** [review]
